### PR TITLE
Enable Partition Discovery for Broker Load

### PR DESCRIPTION
**Pull request by yuanlihan**
Currently, we support loading data from parquet file, but can not parse partition columns in the path of parquet file and can not recursively list all files under base path of input.

This patch is able to discover and infer partitioning information under the base path of input like in Spark. It recursively list all the files under the base path and parse partition columns base on the base path if needed.

This patch parse partition columns at BrokerScanNode.java and save parsing result of each file path as a property of TBrokerRangeDesc, then  parquet_reader of BE can read the value of specified partition column.

### CODE REVIEW

**Comment by imay** [comment]
@yuanlihan 

Firstly, thanks for your improvement. I have some questions about this PR.

1. Should we add an option to switch this feature off? Because this will fill partition row automatically which should be known by users.

2. Can this patch work with other format of file? like CSV format? And how can user use this function.

3. How this function work with other load functions, like user specified columns and set operation. For example, if the partition column name is different in Doris table and Spark.

**Comment by yuanlihan** [comment]
> @yuanlihan
> 
> Firstly, thanks for your improvement. I have some questions about this PR.
> 
> 1. Should we add an option to switch this feature off? Because this will fill partition row automatically which should be known by users.
> 2. Can this patch work with other format of file? like CSV format? And how can user use this function.
> 3. How this function work with other load functions, like user specified columns and set operation. For example, if the partition column name is different in Doris table and Spark.

@imay 
Thanks for your review
1. In my opinion, it is quite intuitive to enable partition discovery as well as recursively list files by default. We extract partition columns iff them was defined/needed in the specified table.
2. We can support partition discovery of other file sources(including Text/CSV/JSON/ORC/Parquet) like in Spark. May be users will prefer having this feature by default, which rarely have conflicts with their previous usage.
3. I will dig more about this and try to solve it.

**Comment by yuanlihan** [comment]
Add Partition Discover as a feature of Broker Load.

For usage of this feature see diffs of related docs.

File: docs/help/Contents/Data Manipulation/broker_load.md [hunk 1]
```diff
@@ -352,15 +357,47 @@
...
-        WITH BROKER hdfs ("username"="hdfs_user", "password"="hdfs_password");
+        WITH BROKER hdfs ("username"="hdfs_user", "password"="hdfs_password"); 
+
+    9. 通过Partition Discovery提取文件路径中的压缩字段
+        如果导入路径为目录，则递归地列出该目录下的所有parquet文件
+        如果需要，则会根据表中定义的字段类型解析文件路径中的partitioned fields，实现类似Spark中读parquet文件
+        1. 不指定Partition Discovery的基础路径（BASE_PATH）
+            LOAD LABEL example_db.label10
+            (
+            DATA INFILE("hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir")
```
Comment (imay) [inline docs/help/Contents/Data Manipulation/broker_load.md]: 这样直接指定目录就可以导入，跟我们之前的默认行为是有冲突的。有的时候用户写错了，可能会将整个目录导入。
为了避免这种情况，可以不使用DATA INFILE，可是使用类似DATA INDIR的方式来指定导入一个目录

File: docs/help/Contents/Data Manipulation/broker_load.md [hunk 2]
```diff
@@ -352,15 +357,47 @@
...
+            则会从文件path中提取k1对应的partitioned field的值，并完成数据导入
+
+        2. 指定Partition Discovery的基础路径（BASE_PATH）
+            LOAD LABEL example_db.label11
+            (
+            DATA INFILE("hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/city=beijing/utc_date=2019-06-26")
+            INTO TABLE `my_table`
+            FORMAT AS "csv"
+            BASE_PATH AS "hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/"
+            (k1, k2, k3, utc_date，city)
```
Comment (imay) [inline docs/help/Contents/Data Manipulation/broker_load.md]: 这里columns表示的一般是文件中包含的列名，如果加上utc_date, city可能会让用户感到困惑。
建议不要再这个地方声明partition列名

File: docs/help/Contents/Data Manipulation/broker_load.md [hunk 3]
```diff
@@ -352,15 +357,47 @@
...
+            hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir目录下包括如下文件：[hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/k1=key1/xxx.parquet, hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/k1=key2/xxx.parquet, ...]
+            则会从文件path中提取k1对应的partitioned field的值，并完成数据导入
+
+        2. 指定Partition Discovery的基础路径（BASE_PATH）
+            LOAD LABEL example_db.label11
+            (
+            DATA INFILE("hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/city=beijing/utc_date=2019-06-26")
+            INTO TABLE `my_table`
+            FORMAT AS "csv"
+            BASE_PATH AS "hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/"
```
Comment (imay) [inline docs/help/Contents/Data Manipulation/broker_load.md]: 是不是可以增加partition column 声明，这样能够让用户清晰的知道自己在做什么。

**Comment by imay** [comment]
> Add Partition Discover as a feature of Broker Load.
> 
> For usage of this feature see diffs of related docs.

@yuanlihan 
Because this patch will change load interface, I think it's better to discuss the new interface before implementation.

**Comment by yuanlihan** [comment]
> > Add Partition Discover as a feature of Broker Load.
> > For usage of this feature see diffs of related docs.
> 
> @yuanlihan
> Because this patch will change load interface, I think it's better to discuss the new interface before implementation.

@imay 
OK. I will create an issue to discuss the interface design about this feature before updating additional diffs.

**Comment by yuanlihan** [comment]
@imay I have added an issue about this [Enable Partition Discovery for Broker Load](https://github.com/apache/incubator-doris/issues/1582)

**Pull request closed without merging**